### PR TITLE
Add integration tests for build script updates, and fix existing bug

### PR DESCRIPTION
**Pull request by jakemac53**
I also broke out a bunch of existing utils in the integration test into a re-usable library, most of the changes here are relating to that.

### CODE REVIEW

File: e2e_example/test/common/utils.dart [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+Directory _toolDir = new Directory(p.join('.dart_tool', 'build'));
+
+Process _process;
+Stream<String> _stdErrLines;
+Stream<String> _stdOutLines;
+
+/// Runs the build script in this package, and waits for the first build to
+/// complete.
+///
+/// By default, this ensures a brand new build by deleting the `.dart_tool`
```
Comment (natebosch) [inline e2e_example/test/common/utils.dart]: [optional] deleting something by default doesn't match typical user behavior or my expectations (even if it's more commonly what we'd do in an integration test). Consider reversing the default.

File: e2e_example/test/common/utils.dart [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    _stdErrLines.listen((line) => print('StdErr: $line'));
+  }
+
+  await Future.wait(extraExpects.map((cb) async => await cb()));
+  await nextSuccessfulBuild;
+}
+
+/// Kills the current build script.
+///
+/// By default this will clean up the `.dart_tool` directory, but you can
```
Comment (natebosch) [inline e2e_example/test/common/utils.dart]: same as above

File: e2e_example/test/common/utils.dart [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      .split('\n')
+      .where((line) => line.startsWith('??'))
+      .map((line) => line.split(' ')[1])
+      .map((path) {
+    var parts = p.split(path);
+    assert(parts[0] == 'e2e_example');
+    return p.joinAll(parts.skip(1));
+  });
+  for (var path in untracked) {
+    Process.runSync('rm', [path]);
```
Comment (natebosch) [inline e2e_example/test/common/utils.dart]: can we use `git clean` command for this?

File: e2e_example/test/common/utils.dart (hunk 1 above)
Comment (jakemac53) [reply e2e_example/test/common/utils.dart]: done

File: e2e_example/test/common/utils.dart (hunk 2 above)
Comment (jakemac53) [reply e2e_example/test/common/utils.dart]: done

File: e2e_example/test/common/utils.dart (hunk 3 above)
Comment (jakemac53) [reply e2e_example/test/common/utils.dart]: ah nice, didn't know about that one, done

**Comment by jakemac53** [comment]
pub.dartlang.org is down which is causing the travis failure, it passes locally though so I am going to go ahead and submit this.

**Pull request merged**